Not able to import /async, /creatable, .etc from an ES module

### [Minimal repro](https://github.com/srmagura/react-select-esm-repro)

If placed in an `.mjs` file, the following code results in an error.

```js
import AsyncSelect from "react-select/async";

console.log(AsyncSelect);
```

```text
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import 'C:\Projects\OSS\Bugs\react-select-esm-repro\node_modules\react-select\async' is not supported resolving ES modules imported from C:\Projects\OSS\Bugs\react-select-esm-repro\index.mjs
Did you mean to import react-select-esm-repro/node_modules/react-select/async/dist/react-select.cjs.js?
    at finalizeResolution (internal/modules/esm/resolve.js:281:17)
    at moduleResolve (internal/modules/esm/resolve.js:708:10)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:819:11)
    at Loader.resolve (internal/modules/esm/loader.js:89:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:242:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:73:40)
    at link (internal/modules/esm/module_job.js:72:36) {
  code: 'ERR_UNSUPPORTED_DIR_IMPORT',
  url: 'file:///C:/Projects/OSS/Bugs/react-select-esm-repro/node_modules/react-select/async'
}
```

While this reproduction uses Node, the issue also occurs when using Webpack 5 which follows Node's handling of ES module imports. Here's the Webpack error for Googleability:

```text
Module not found: Error: Can't resolve 'react-select/async' in 'C:\Projects\ITI\iti-react\iti-react\dist'
Did you mean 'react-select.esm.js'?
BREAKING CHANGE: The request 'react-select/async' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
```

## Suggested Solution

Utilize the `exports` field in `package.json` [as described in the Node documentation.](https://nodejs.org/api/packages.html#packages_subpath_exports)

## Additional Context

- react-select version: 5.1.0
- Node version: 14.17.6

--- a/.changeset/nervous-islands-shake.md
+++ b/.changeset/nervous-islands-shake.md
@@ -1,0 +1,5 @@
+---
+'react-select': patch
+---
+
+Fix importing `react-select` in Node ESM

--- a/docs/generate-magical-types/generate/package.json
+++ b/docs/generate-magical-types/generate/package.json
@@ -1,3 +1,4 @@
 {
-  "main": "dist/react-select-generate-magical-types-generate.cjs.js"
+  "main": "dist/react-select-generate-magical-types-generate.cjs.js",
+  "module": "dist/react-select-generate-magical-types-generate.esm.js"
 }

--- a/docs/generate-magical-types/package.json
+++ b/docs/generate-magical-types/package.json
@@ -1,6 +1,19 @@
 {
   "name": "@react-select/generate-magical-types",
   "main": "dist/generate-magical-types.cjs.js",
+  "exports": {
+    "./generate": {
+      "module": "./generate/dist/react-select-generate-magical-types-generate.esm.js",
+      "import": "./generate/dist/react-select-generate-magical-types-generate.cjs.mjs",
+      "default": "./generate/dist/react-select-generate-magical-types-generate.cjs.js"
+    },
+    "./serialize": {
+      "module": "./serialize/dist/react-select-generate-magical-types-serialize.esm.js",
+      "import": "./serialize/dist/react-select-generate-magical-types-serialize.cjs.mjs",
+      "default": "./serialize/dist/react-select-generate-magical-types-serialize.cjs.js"
+    },
+    "./package.json": "./package.json"
+  },
   "//": "these deps aren't real, they're just to appease preconstruct",
   "dependencies": {
     "@babel/runtime": "*",

--- a/docs/generate-magical-types/serialize/package.json
+++ b/docs/generate-magical-types/serialize/package.json
@@ -1,3 +1,4 @@
 {
-  "main": "dist/react-select-generate-magical-types-serialize.cjs.js"
+  "main": "dist/react-select-generate-magical-types-serialize.cjs.js",
+  "module": "dist/react-select-generate-magical-types-serialize.esm.js"
 }

--- a/package.json
+++ b/package.json
@@ -24,7 +24,7 @@
     "@emotion/babel-plugin": "^11.10.2",
     "@emotion/jest": "^11.1.0",
     "@manypkg/cli": "^0.19.2",
-    "@preconstruct/cli": "^2.6.0",
+    "@preconstruct/cli": "^2.6.2",
     "@testing-library/dom": "8.19.0",
     "@testing-library/jest-dom": "5.1.1",
     "@testing-library/react": "12.1.4",

--- a/package.json
+++ b/package.json
@@ -24,7 +24,7 @@
     "@emotion/babel-plugin": "^11.10.2",
     "@emotion/jest": "^11.1.0",
     "@manypkg/cli": "^0.19.2",
-    "@preconstruct/cli": "^2.2.2",
+    "@preconstruct/cli": "^2.6.0",
     "@testing-library/dom": "8.19.0",
     "@testing-library/jest-dom": "5.1.1",
     "@testing-library/react": "12.1.4",
@@ -120,11 +120,10 @@
     "packages": [
       "packages/*",
       "docs/generate-magical-types"
-    ]
-  },
-  "browser": {
-    "./dist/react-select.cjs.js": "./dist/react-select.browser.cjs.js",
-    "./dist/react-select.esm.js": "./dist/react-select.browser.esm.js"
+    ],
+    "exports": {
+      "importConditionDefaultExport": "default"
+    }
   },
   "resolutions": {
     "csstype": "^3.0.2"

--- a/packages/react-select/package.json
+++ b/packages/react-select/package.json
@@ -62,34 +62,35 @@
   },
   "exports": {
     ".": {
-      "require": "./dist/react-select.cjs.js",
-      "import": "./dist/react-select.esm.js",
-      "types": "./dist/react-select.cjs.d.ts"
+      "module": "./dist/react-select.esm.js",
+      "import": "./dist/react-select.cjs.mjs",
+      "default": "./dist/react-select.cjs.js"
     },
     "./base": {
-      "require": "./base/dist/react-select-base.cjs.js",
-      "import": "./base/dist/react-select-base.esm.js",
-      "types": "./base/dist/react-select-base.cjs.d.ts"
-    },
-    "./animated": {
-      "require": "./animated/dist/react-select-animated.cjs.js",
-      "import": "./animated/dist/react-select-animated.esm.js",
-      "types": "./animated/dist/react-select-animated.cjs.d.ts"
+      "module": "./base/dist/react-select-base.esm.js",
+      "import": "./base/dist/react-select-base.cjs.mjs",
+      "default": "./base/dist/react-select-base.cjs.js"
     },
     "./async": {
-      "require": "./async/dist/react-select-async.cjs.js",
-      "import": "./async/dist/react-select-async.esm.js",
-      "types": "./async/dist/react-select-async.cjs.d.ts"
+      "module": "./async/dist/react-select-async.esm.js",
+      "import": "./async/dist/react-select-async.cjs.mjs",
+      "default": "./async/dist/react-select-async.cjs.js"
+    },
+    "./animated": {
+      "module": "./animated/dist/react-select-animated.esm.js",
+      "import": "./animated/dist/react-select-animated.cjs.mjs",
+      "default": "./animated/dist/react-select-animated.cjs.js"
     },
     "./creatable": {
-      "require": "./creatable/dist/react-select-creatable.cjs.js",
-      "import": "./creatable/dist/react-select-creatable.esm.js",
-      "types": "./creatable/dist/react-select-creatable.cjs.d.ts"
+      "module": "./creatable/dist/react-select-creatable.esm.js",
+      "import": "./creatable/dist/react-select-creatable.cjs.mjs",
+      "default": "./creatable/dist/react-select-creatable.cjs.js"
     },
     "./async-creatable": {
-      "require": "./async-creatable/dist/react-select-async-creatable.cjs.js",
-      "import": "./async-creatable/dist/react-select-async-creatable.esm.js",
-      "types": "./async-creatable/dist/react-select-async-creatable.cjs.d.ts"
-    }
+      "module": "./async-creatable/dist/react-select-async-creatable.esm.js",
+      "import": "./async-creatable/dist/react-select-async-creatable.cjs.mjs",
+      "default": "./async-creatable/dist/react-select-async-creatable.cjs.js"
+    },
+    "./package.json": "./package.json"
   }
 }

